apis/v1/httproute_types.go

```diff
@@ -196,8 +196,19 @@ type HTTPRouteRule struct {
 	// Filters define the filters that are applied to requests that match
 	// this rule.
 	//
-	// The effects of ordering of multiple behaviors are currently unspecified.
-	// This can change in the future based on feedback during the alpha stage.
+	// Wherever possible, implementations SHOULD implement filters in the order
```

Comment:
Id prefer if we removed `Wherever possible` and instead
* made ordering for `CORE` filters a `MUST`
* made ordering for `EXTENDED` and `IMPLEMENTATION SPECIFIC` filters a `SHOULD`

should be easy to a conformance test for `CORE` by adding combinations of - `set`, `add`, & `remove` settings of `requestHeaderModifier` filter

Reply:
I don't understand why the ordering requirement is linked to CORE vs extended/implementation specific filters. Also if the filters list contains, say, both core and extended in a specific order the MUST requirement only applies to the core filters subset (or subsequence)?

Reply:
It looks like we consider RequestHeaderModifier and RequestRedirect as our core filters and the rest are extended:

https://github.com/kubernetes-sigs/gateway-api/blob/8e62cc338e69bec9cb9fd01cf59af421e3aba879/apis/v1/httproute_types.go#L748-L799

If we _required_ implementations to have a strict interpretation of these filters, that would mean that RequestHeaderModifier would always have to be specified before RequestRedirect. Although this is probably reasonable in practice, it could end up breaking existing users which would not be good.

We do actually have validation that prevents the RequestHeaderModifier filter from being repeated though: https://github.com/kubernetes-sigs/gateway-api/blob/8e62cc338e69bec9cb9fd01cf59af421e3aba879/apis/v1/httproute_types.go#L225

I'm hesitant to add a new _requirement_ to the spec here. This could complicate adding future filters, but the more problematic aspect is that it could invalidate some existing configuration.

My rationale for suggesting `Wherever possible` is that I wanted to clearly recognize that many implementations simply cannot adjust the order of when certain filters are executed, I think that still applies here.

Reply:
I was trying to find wording that can be converted into conformance tests :), but this works, to an implementor and user, this reads as "best effort", which sgtm

Reply:
Yeah at least with the existing set of "core" filters, I'm not sure how much utility there would be in enforcing ordering for these filters, but maybe down the line depending on what is promoted to core. If we do intend to make respecting ordering for core filters a requirement, we should add a conformance test, e.g. if response header modification becomes core then asserting a redirect response does/does not have certain modified headers depending if the filters are in the right order.

My understanding of https://github.com/kubernetes-sigs/gateway-api/issues/1598 was that we just wanted to add some content to the spec saying that the list of filters might be interpreted as ordered, depending on the implementation. I might get rid of the `SHOULD` sentence that starts this diff and start off the change with something like `Implementations MAY choose to implement filters in the order they are specified. Such implementations MAY choose to implement this ordering strictly...`

Reply:
I could go either way on the "wherever possible ... SHOULD" wording, if the intent is to intentionally be prescriptive then it makes sense, I think there are other instances of SHOULD being used to that end in the spec

Reply:
@robscott and others what's the final decision here? If the existing wording is okay, can we go ahead with approval and merging etc? Thanks

Reply:
I think that the SHOULD already implies the "wherever possible" - in this case it's actually making the intent more clear. I think this wording is good to go.